virtio-devices/src/iommu.rs

```diff
@@ -563,15 +572,25 @@ impl Request {
                         }
                     }
 
-                    // Remove mapping associated with the domain
-                    mapping
-                        .domains
-                        .write()
-                        .unwrap()
-                        .get_mut(&domain_id)
-                        .unwrap()
-                        .mappings
-                        .remove(&virt_start);
+                    // Remove all mappings associated with the domain
+                    let mut size = req.virt_end - virt_start + 1;
+                    let mut virt_start = req.virt_start;
+                    while size > 0 {
+                        if let Some(addr_map) = mapping
+                            .domains
+                            .write()
+                            .unwrap()
+                            .get_mut(&domain_id)
+                            .unwrap()
+                            .mappings
+                            .remove(&virt_start)
+                        {
+                            size -= addr_map.size;
```

Comment:
The virtio-iommu spec explicitly specifies that partial unmappings are not supported. See [this](https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.html#x1-5180002) section of the specification.

```
If a mapping affected by the range is not covered in its entirety by the range (the UNMAP request would split the mapping), then the device SHOULD set the request status to VIRTIO_IOMMU_S_RANGE, and SHOULD NOT remove any mapping.
```